### PR TITLE
Fix GNOME gir generation with lists of dependency lists

### PR DESCRIPTION
**Pull request by nirbheek**
Also add a repr() for all objects used by the interpreter, which is needed because in some places we include them in logs and the output is totally unusable otherwise.

### CODE REVIEW

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

File: run_unittests.py [hunk 1]
```diff
@@ -398,6 +398,33 @@ def test_needs_exe_wrapper_override(self):
...
+        self.assertEqual([holder1, 2, 3], listify([holder1, 2, [3]]))
+        self.assertEqual([1], listify(holder1, unholder=True))
+        self.assertEqual([1], listify([holder1], unholder=True))
+        self.assertEqual([1, 2], listify([holder1, 2], unholder=True))
+        self.assertEqual([1, 2, 3], listify([holder1, 2, [holder3]], unholder=True))
+        # Unholding doesn't work recursively when not flattening
+        self.assertEqual([1, [2], [holder3]], listify([holder1, [2], [holder3]], unholder=True, flatten=False))
+
+    def test_extract_as_list(self):
+        extract = mesonbuild.mesonlib.extract_as_list
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [F841] local variable 'extract' is assigned to but never used

File: mesonbuild/modules/gnome.py [hunk 2]
```diff
@@ -323,11 +323,9 @@ def _get_dependencies_flags(self, deps, state, depends=None, include_rpath=False
         cflags = OrderedSet()
         ldflags = OrderedSet()
         gi_includes = OrderedSet()
-        deps = mesonlib.listify(deps)
+        deps = listify(deps, unholder=True)
```
Comment (ignatenkobrain) [inline mesonbuild/modules/gnome.py]: __[flake8]__
* [F821] undefined name 'listify'

File: mesonbuild/modules/gnome.py [hunk 3]
```diff
@@ -525,9 +523,8 @@ def generate_gir(self, state, args, kwargs):
             else:
                 raise MesonException('Gir export packages must be str or list')
 
-        deps = mesonlib.extract_as_list(kwargs, 'dependencies', pop = True)
         deps = (girtarget.get_all_link_deps() + girtarget.get_external_deps() +
-                deps)
+                extract_as_list(kwargs, 'dependencies', pop=True, unholder=True))
```
Comment (ignatenkobrain) [inline mesonbuild/modules/gnome.py]: __[flake8]__
* [F821] undefined name 'extract_as_list'

File: mesonbuild/modules/gnome.py [hunk 4]
```diff
@@ -794,7 +791,8 @@ def gtkdoc(self, state, args, kwargs):
 
     def _get_build_args(self, kwargs, state):
         args = []
-        cflags, ldflags, gi_includes = self._get_dependencies_flags(kwargs.get('dependencies', []), state, include_rpath=True)
+        deps = extract_as_list(kwargs, 'dependencies', unholder=True)
```
Comment (ignatenkobrain) [inline mesonbuild/modules/gnome.py]: __[flake8]__
* [F821] undefined name 'extract_as_list'

**Comment by nirbheek** [comment]
Fixed, and I also added a fix for https://github.com/mesonbuild/meson/issues/2408 and many other instances of that same thing.

**Comment by nirbheek** [comment]
Seems to be all good now. Also contains some tests for `listify` etc.

**Pull request merged**